### DISCUSSION TITLE
Custom components and Hosting extensions.

### DISCUSSION THREAD

**Discussion by peteraritchie** [microsoft/aspire#1243] [Q&A] [answered]
From what I've read so far, a "Component" focuses on resource project integration: supported by VS tooling _Add/.NET Aspire Component_, `IHostApplicationBuilder` extensions, optionally `IApplicationBuilder` extensions, configuration schema and type, etc.; published in a NuGet package.  Depending on the component (?), there's an expectation that a consumer will include the component in the app model through extensions to `IDistributedApplicationBuilder`.  I wouldn't expect those extensions to live in the NuGet package that a consumer would install via _Add/.NET Aspire Component_, but be part of developing/publishing a .NET Aspire Component.

[AspireWithNode](https://github.com/dotnet/aspire-samples/tree/main/samples/AspireWithNode) gets into what `IDistributedApplicationBuilder` extensions might look like, but not how but not how to publish them as a component author.

Is it expected that a component author publishes orchestrator extensions via another NuGet package, and are there any guidance/expectations/rules I might have missed?

**Comment by DamianEdwards** [comment, marked as answer]
> Is it expected that a component author publishes orchestrator extensions via another NuGet package, and are there any guidance/expectations/rules I might have missed?

In short, yes. "Components" are the packages you integrate into individual service apps, and many of those will have corresponding hosting extensions packages for composing instances of the service being connected to into the AppHost as resources. They're related but separate.
